truelearn/datasets/_base.py

```diff
@@ -0,0 +1,105 @@
+import dataclasses
+import hashlib
+from typing import Optional
+from urllib import request
+from os import path
+
+
+# pylint: disable=pointless-string-statement
+"""
+Copyright of `RemoteFileMetaData`, `download_file` are held by
+[BSD 3-Clause License, scikit-learn developers, 2007-2022].
+"""
+
+
+@dataclasses.dataclass
+class RemoteFileMetaData:
+    """Remote file metadata.
+
+    Args:
+        url: The url of the file.
+        filename: The filename of the downloaded file.
+        sha256_expected: The expected sha256 sum of the file.
+    """
+
+    url: str
+    filename: str
+    sha256: str
+
+
+def _sha256sum(filepath) -> str:
+    buf_size = 65536
+    h = hashlib.sha256()
+    with open(filepath, "rb", buffering=0) as file:
+        while True:
+            chunk = file.read(buf_size)
+            if not chunk:
+                break
+            h.update(chunk)
+    return h.hexdigest()
+
+
+def _download_file(*, filepath: str, url: str, sha256sum: str) -> None:
+    """Download a remote file and check the sha256.
+
+    Args:
+        filepath:
+            The full path of the created file.
+        url:
+            The url of the file.
+        sha256sum:
+            The expected sha256sum of the file.
+    """
+    if url.lower().startswith("http"):
+        print(f"Downloading {url} into {filepath}")
+        # the bandit warning is suppressed here
+        # because we have checked whether the url starts with http
+        request.urlretrieve(url, filepath)  # nosec
```

Comment:
Is there a reason we allow http, instead of only https? nosec means no security i assume

Reply:
This also allows `https` because they start with the same prefix `http`.

Reply:
> Is there a reason we allow http, instead of only https? nosec means no security i assume

`nosec` is designed to suppress bandit's warnings about using `urlretrieve`. Even if you check that the url is the intended one, bandit will still issue this warning.